### ISSUE TITLE
Loaded enemy NPCs sometimes fail to assume defensive cover positions

### ISSUE THREAD

**Issue by Decane** [Decane/SRP#123]
### The Problem:

Enemy NPCs stationed at camps are supposed to assume defensive cover positions when they spot the player approaching. This is implemented via `CCover_manager:calculate_covers(position)` in cover_manager.script for squads in CCover_manager member table `self.squads`, of which there exists one per smart terrain. Alas, GSC forgot to persist this table across saves - the save/load functions in cover_manager.script ignore it.

This means _loaded_ enemy NPCs will sometimes not congregate into a defensive formation when one of them sees the player approaching their camp.

### Recent Regression:

Before https://github.com/Decane/SRP/commit/40df7c62514f6e6f923e87541550b71878d08d6d, this bug was mitigated by the fact that every so often, 'stayer' squads would be unregistered from, and re-registered to, their smart terrain. The code specifically responsible for the mitigation was the following in `sim_stay_point:is_finished()` in sim_squad_generic.script:
```
	if squad.sim_combat_id ~= nil then
		src_smart.combat_manager:unregister_squad(squad)
	end
```
The reason this enabled NPCs to assume defensive cover positions correctly is that src_smart.combat_manager stores a CCombat_manager instance, and calling `CCombat_manager:unregister_squad(squad)` -- which the code above hence did -- enables a subsequent `CCombat_manager:register_squad(squad)` call from `sim_stay_point:make()` to reach `CCover_manager:register_squad(squad)`, which in turn adds the relevant squad to the relevant cover manager instance's `self.squads` table. It also sets flag `self.is_invalid = false`. This flag 'unblocks' `CCover_manager:calculate_covers(position)`, allowing it to set the cover positions for squads belonging to the cover manager instance's `self.squads` table.

In commit https://github.com/Decane/SRP/commit/40df7c62514f6e6f923e87541550b71878d08d6d however, I commented out the above code with the following justification:

>Not necessary for stay -> stay transitions; done in sim_attack_point:make for stay -> attack transitions

This had the consequence of blocking the code path leading from `sim_stay_point:make()` to `CCover_manager:register_squad(squad)`, which in turn meant that calls to `CCover_manager:calculate_covers(position)` would no longer make NPCs assume defensive positions.

### Candidate Solution:

The ideal fix would be to address the root cause and add the following to `CCover_manager:save(packet)` and `CCover_manager:load(packet)`, respectively:
```
	n = 0
	for k in pairs (self.squads) do
		n = n + 1
	end
	packet:w_u8(n)
	for k in pairs (self.squads) do
		packet:w_stringZ(k)
	end
```
```
	n = packet:r_u8()
	for i = 1, n do
		local squad_id = packet:r_stringZ()
		self.squads[squad_id] = true
	end
```
However, it seems we can't do that without modifying the save data structure expected by the engine, because the aforementioned functions are called from STATE_Write / STATE_Read functions, as follows:

- `se_smart_terrain:STATE_Write(packet)` [smart_terrain.script] calls ...
  - `CCombat_manager:save(packet)` [sim_combat.script] calls ...
    - `CCover_manager:save(packet)` [cover_manager.script]
- `se_smart_terrain:STATE_Read(packet, size)` [smart_terrain.script] calls ...
  - `CCombat_manager:load(packet)` [sim_combat.script] calls ...
    - `CCover_manager:load(packet)` [cover_manager.script]

If we do make the suggested changes, the game will crash on load / new game:
```
! [LUA][ERROR] Abort(): [_G.set_save_marker]: Save marker mismatch '_CCover_manager' 0 ~= 6.
 
stack traceback:
	_g.script:19: in function 'abort'
	_g.script:456: in function 'set_save_marker'
	cover_manager.script:107: in function 'load'
	sim_combat.script:837: in function 'load'
	smart_terrain.script:630: in function <smart_terrain.script:616>
```

### Additional Caveat:

Even if the fix above were viable, at least two further changes would be necessary to compensate the loss of `self.is_invalid = false` being periodically set by stay -> stay transitions:
```diff
diff --git a/gamedata/scripts/sim_combat.script b/gamedata/scripts/sim_combat.script
index b366e5a..02c4058 100644
--- a/gamedata/scripts/sim_combat.script
+++ b/gamedata/scripts/sim_combat.script
@@ -188,6 +188,7 @@ function CCombat_manager:enemy_callback(my_squad, enemy_squad)
 --                                     if not data._attack_squad then
                                                data._stage = "defence"
                                                data._inited_defend_time = game.get_game_time()
+                                               self.cover_manager.is_invalid = false
 --                                     else
 --                                             dbgmsg("Rejected idle --> defence switch for squad "..k.." because it is attacking.")
 --                                     end
@@ -361,6 +362,7 @@ function CCombat_manager:update_actor()
                                if data._stage == "idle" then
                                        data._stage = "defence"
                                        data._inited_defend_time = game_time
+                                       self.cover_manager.is_invalid = false
                                end
                        end
                        -- DC170501:
```

### Why Not Revert?

I would like to avoid reinstating the commented code in `sim_stay_point:is_finished()` because it was never a very good solution to begin with; it still left various circumstances where NPCs would not congregate into defensive cover positions when they should.

**Comment by Decane** [comment]
Fixed by https://github.com/Decane/SRP/commit/482cec2c14f34cb1cdc5ca38820ad2d5528e51b6, with a few caveats:

1. Occasionally, `npc_obj:find_best_cover()` returns nil when `CCover_manager:calculate_covers()` is called by `xr_gulag.defending_smart()`. This leaves the affected NPC(s) with their non-cover job (usually a patrol job or squad leader job) even while the rest of their squad takes a defensive formation facing the approaching enemy squad. This is a separate issue from what is fixed in this commit and can sometimes be worked around with the following tweak:
```diff
	if cover ~= nil then
		...
+	else
+		self.is_invalid = false
	end
```
Essentially, this tweak just enables `CCover_manager:calculate_covers()` to get called over and over again until cover is assigned to all NPCs. Unfortunately, it doesn't always succeed, even after thousands of attempts.

2. The reason why `xr_gulag.defending_smart()` doesn't need an equivalent of sim_combat.script's `self.cover_manager.is_invalid = false` before calling `CCover_manager:calculate_covers()` is that these calls only happen during the counter-attack phase of a 'capture' task, which can only be active if the defending squad was just recently registered as a stayer squad to the combat manager of the relevant smart terrain. And this would have involved calling that combat manager's cover manager's `CCover_manager:register_squad()` method to set `self.is_current = false`.
